decoder.go

```diff
@@ -146,60 +179,71 @@ func (d *decoder) decode(nalu []byte) (image.Image, error) {
 	avPacket.size = C.int(len(nalu))
 	res := C.avcodec_send_packet(d.codecCtx, &avPacket)
 	if res < 0 {
+		//nolint:nilnil // avcodec_send_packet raises non-critical errors
 		return nil, nil
 	}
 
 	// receive frame if available
-	res = C.avcodec_receive_frame(d.codecCtx, d.srcFrame)
+	res = C.avcodec_receive_frame(d.codecCtx, d.src.frame)
 	if res < 0 {
+		//nolint:nilnil // avcodec_receive_frame raises non-critical errors
 		return nil, nil
 	}
 
-	// if frame size has changed, allocate needed objects
-	if d.dstFrame == nil || d.dstFrame.width != d.srcFrame.width || d.dstFrame.height != d.srcFrame.height {
-		if d.dstFrame != nil {
-			C.av_frame_free(&d.dstFrame)
-		}
+	// get a frame from the pool
+	d.dst = d.avFramePool.Get().(*avFrameWrapper)
+	if d.dst == nil {
+		return nil, errors.New("failed to obtain AVFrame from pool")
+	}
 
+	// if frame size has changed, reinitialize dstFrame
+	if d.dst.frame.width != d.src.frame.width || d.dst.frame.height != d.src.frame.height {
 		if d.swsCtx != nil {
 			C.sws_freeContext(d.swsCtx)
 		}
 
-		d.dstFrame = C.av_frame_alloc()
-		d.dstFrame.format = C.AV_PIX_FMT_RGBA
-		d.dstFrame.width = d.srcFrame.width
-		d.dstFrame.height = d.srcFrame.height
-		d.dstFrame.color_range = C.AVCOL_RANGE_JPEG
-		res = C.av_frame_get_buffer(d.dstFrame, 1)
+		dstFrame, err := allocateAVFrame()
```

Comment:
Why do we need to allocate here if we already got a frame from the pool here? https://github.com/erh/viamrtsp/pull/41/files#diff-2a967f84f1af0d52111b8dbfbeadc44bfeb1092dacebb2bc0648a0e57284c04bR194

Reply:
I think Dan's notes help a lot. Basically the frame we got from the pool may be a fresh blank frame with no metadata, or if the stream settings changed, a frame with the wrong metadata. That's why we reallocate here